### PR TITLE
feat: add URL dependencies support to consume shared module via module federation

### PR DESCRIPTION
**Pull request by snitin315**
Fixes https://github.com/webpack/webpack/pull/14035

> For module federation shared, without `requiredVersion` configured, it will infer the version based on package.json, as we can see in ConsumeSharedPlugin.js.
>
> But when we get `requiredVersion` from `package.json`, it just gets dep/dev /peer/optional dependencies value, without consideration of some URL Dependencies cases.
>
> For example, `"isarray": "git+https://github.com/juliangruber/isarray.git#v2.0.3"`, it's real version should be `v2.0.3`, but we can not get that.

### <samp>🤖 Generated by Copilot at fb959e1</samp>

This pull request adds support for git urls as dependencies in shared modules. It implements a new function `normalizeVersion` to handle git urls in `lib/sharing/utils.js` and adds more tests and dependencies to the `test/configCases/sharing/consume-multiple-versions` folder to verify the feature. It also updates the `cspell.json` file to ignore the word "commithash".

### <samp>🤖 Generated by Copilot at fb959e1</samp>

*  Add `normalizeVersion` function to `lib/sharing/utils.js` to normalize the version of a shared module from a git url ([link](https://github.com/webpack/webpack/pull/16945/files?diff=unified&w=0#diff-260b12729ec74ae844ab979b5bb4344874cd0a89eeb39dc78a66d284a44a2919L12-R303))
*  Use `normalizeVersion` function on the optional, dependencies, peer dependencies, and dev dependencies of a shared module in `getRequiredVersion` function in `lib/sharing/utils.js` to ensure consistent version resolution ([link](https://github.com/webpack/webpack/pull/16945/files?diff=unified&w=0#diff-260b12729ec74ae844ab979b5bb4344874cd0a89eeb39dc78a66d284a44a2919L67-R350), [link](https://github.com/webpack/webpack/pull/16945/files?diff=unified&w=0#diff-260b12729ec74ae844ab979b5bb4344874cd0a89eeb39dc78a66d284a44a2919L74-R357), [link](https://github.com/webpack/webpack/pull/16945/files?diff=unified&w=0#diff-260b12729ec74ae844ab979b5bb4344874cd0a89eeb39dc78a66d284a44a2919L81-R364), [link](https://github.com/webpack/webpack/pull/16945/files?diff=unified&w=0#diff-260b12729ec74ae844ab979b5bb4344874cd0a89eeb39dc78a66d284a44a2919L88-R371))
*  Add `commithash` to the cspell ignore list in `cspell.json` to avoid spelling errors in the code that deals with git commit hashes ([link](https://github.com/webpack/webpack/pull/16945/files?diff=unified&w=0#diff-77519f787fe7e051a14c588f101d61e4d8f0d3b01bf8e16cb65d3838243d4e68L280-R281))
*  Add more shared modules with different versions and git urls as dependencies to the `test/configCases/sharing/consume-multiple-versions` test case to test the new feature of allowing git urls as dependencies in the package.json of a shared module ([link](https://github.com/webpack/webpack/pull/16945/files?diff=unified&w=0#diff-09763a2cacca4a42cceef20059e1c38bd8612813092e8a4f7b2b1edb65fd693fR36-R148), [link](https://github.com/webpack/webpack/pull/16945/files?diff=unified&w=0#diff-4f15508012dc80291774582d9ea17dbd78b35feb113dbdd71b8c4eaa91fb91d0L3-R25), [link](https://github.com/webpack/webpack/pull/16945/files?diff=unified&w=0#diff-9ce88b51918be9a8f7baf725de182e8c677103c7d4f325bc9c0bb0c2ba38d7f1R15-R90))
*  Add more expectations to the `test/configCases/sharing/consume-multiple-versions/index.js` file to check the versions of the shared modules that are required by the host module ([link](https://github.com/webpack/webpack/pull/16945/files?diff=unified&w=0#diff-09763a2cacca4a42cceef20059e1c38bd8612813092e8a4f7b2b1edb65fd693fR164-R188))

### CODE REVIEW

File: lib/sharing/utils.js [hunk 1]
```diff
@@ -9,13 +9,296 @@ const { join, dirname, readJson } = require("../util/fs");
...
+	return getVersionFromHash(gitUrl);
+}
+
 /**
  * @param {string} str maybe required version
  * @returns {boolean} true, if it looks like a version
  */
-exports.isRequiredVersion = str => {
+function isRequiredVersion(str) {
 	return /^([\d^=v<>~]|[*xX]$)/.test(str);
```
Comment (TheLarkInn) [inline lib/sharing/utils.js:280]: This isn't a stateful regex so hoist it to top of file like the rest of them and give a helpful name.

**Review by TheLarkInn — CHANGES_REQUESTED** [review]

**Comment by alexander-akait** [comment]
@ScriptedAlchemy Friendly ping, what do you think about it?

File: lib/sharing/utils.js [hunk 2]
```diff
@@ -9,13 +9,299 @@ const { join, dirname, readJson } = require("../util/fs");
 
 /** @typedef {import("../util/fs").InputFileSystem} InputFileSystem */
 
+// Extreme shorthand only for github. eg: foo/bar
+const RE_URL_GITHUB_EXTREME_SHORT = /^[^/@:.\s][^/@:\s]*\/[^@:\s]*[^/@:\s]#\S+/;
+
+// Short url with specific protocol. eg: github:foo/bar
+const RE_GIT_URL_SHORT = /^(github|gitlab|bitbucket|gist):\/?[^/.]+\/?/;
```
Comment (alexander-akait) [inline lib/sharing/utils.js:16]: let's add `i`, because https://GITHUB.COM/, hope no one write it in this way

File: lib/sharing/utils.js [hunk 3]
```diff
@@ -9,13 +9,299 @@ const { join, dirname, readJson } = require("../util/fs");
...
+
+// Currently supported protocols
+const RE_PROTOCOL =
+	/^((git\+)?(ssh|https?|file)|git|github|gitlab|bitbucket|gist):$/;
```
Comment (alexander-akait) [inline lib/sharing/utils.js:20]: Same here, let's add `i`

File: lib/sharing/utils.js [hunk 4]
```diff
@@ -9,13 +9,299 @@ const { join, dirname, readJson } = require("../util/fs");
...
+
+// Has custom protocol
+const RE_CUSTOM_PROTOCOL = /^((git\+)?(ssh|https?|file)|git):\/\//;
```
Comment (alexander-akait) [inline lib/sharing/utils.js:23]: And here

File: lib/sharing/utils.js [hunk 5]
```diff
@@ -9,13 +9,299 @@ const { join, dirname, readJson } = require("../util/fs");
...
+const PROTOCOLS_FOR_SHORT = [
+	"github:",
+	"gitlab:",
+	"bitbucket:",
+	"gist:",
+	"file:"
+];
+
+// Default protocol for git url
+const DEF_GIT_PROTOCOL = "git+ssh://";
```
Comment (alexander-akait) [inline lib/sharing/utils.js:51]: And here

**Review by alexander-akait — CHANGES_REQUESTED** [review]
@snitin315 There are many places where we need to normalize or check UPPERCASE versions too, so let's fix all of them and we can merge it

File: lib/sharing/utils.js (hunk 5 above)
Comment (snitin315) [reply lib/sharing/utils.js:51]: This is a string.

File: lib/sharing/utils.js [hunk 6]
```diff
@@ -9,13 +9,299 @@ const { join, dirname, readJson } = require("../util/fs");
...
+const VERSION_PATTERN_REGEXP = /^([\d^=v<>~]|[*xX]$)/;
+
+// Specific protocol for short url without normal hostname
+const PROTOCOLS_FOR_SHORT = [
+	"github:",
+	"gitlab:",
+	"bitbucket:",
+	"gist:",
+	"file:"
+];
```
[the remark this answers is not in the document]
Comment (snitin315) [reply lib/sharing/utils.js:48]: Yeah, updated the check to use `PROTOCOLS_FOR_SHORT.includes(protocol.toLowerCase())`

**Review by alexander-akait — APPROVED** [review]
Thank you

**Review by TheLarkInn — APPROVED** [review]
:shipit:

**Comment by TheLarkInn** [comment]
@snitin315 @chenxsan Let's make sure we document the specific feature scenarios here.

**Pull request merged**